services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightWorkManager.java

```diff
@@ -158,6 +158,61 @@ public void getCatalogs(FlightProducer.ServerStreamListener listener, BufferAllo
     }
   }
 
+  /**
+   * Submits a GET_SCHEMAS job to a worker and sends the response to given ServerStreamListener.
+   *
+   * @param listener    ServerStreamListener listening to the job result.
+   * @param allocator   BufferAllocator used to allocate the response VectorSchemaRoot.
+   * @param userSession The session for the user which made the request.
+   */
+  public void getSchemas(String catalog, String schemaFilterPattern,
+                         FlightProducer.ServerStreamListener listener,
+                         BufferAllocator allocator,
+                         Supplier<Boolean> isRequestCancelled,
+                         UserSession userSession) {
+    final UserBitShared.ExternalId runExternalId = ExternalIdHelper.generateExternalId();
+
+    UserProtos.GetSchemasReq.Builder reqBuilder = UserProtos.GetSchemasReq.newBuilder();
+    if (catalog != null) {
+      UserProtos.LikeFilter filter = UserProtos.LikeFilter.newBuilder().setPattern(catalog).build();
+      reqBuilder.setCatalogNameFilter(filter);
+    }
+    if (schemaFilterPattern != null) {
+      UserProtos.LikeFilter filter = UserProtos.LikeFilter.newBuilder().setPattern(schemaFilterPattern).build();
+      reqBuilder.setSchemaNameFilter(filter);
+    }
+
+    final UserRequest userRequest = new UserRequest(UserProtos.RpcType.GET_SCHEMAS, reqBuilder.build());
```

Comment:
How much of this and below is common between this and getCatalogs? Can it be refactored?

Reply:
Not much, since they are different schemas and different response objects. We couldn't think of a good way to refactor it. Do you have any suggestions?

Reply:
This is fine then.